source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/CapabilitiesServiceV2DecoratorBuilder.cs

```diff
@@ -44,23 +39,47 @@ public CapabilitiesServiceV2DecoratorBuilder DecorateGetCapabilitiesWith(GetCapa
 
         public Decorator<IAsyncClientCapabilitiesServiceV2> Build()
         {
-            return inner => new FuncCapabilitiesServiceV2Decorator(inner, getCapabilitiesFunc);
+            return inner => new FuncCapabilitiesServiceV2Decorator(
+                inner,
+                getCapabilitiesFunc,
+                beforeGetCapabilities,
+                afterGetCapabilities
+            );
         }
 
-        private class FuncCapabilitiesServiceV2Decorator : IAsyncClientCapabilitiesServiceV2
+        class FuncCapabilitiesServiceV2Decorator : IAsyncClientCapabilitiesServiceV2
         {
-            private readonly IAsyncClientCapabilitiesServiceV2 inner;
-            private readonly GetCapabilitiesClientDecorator getCapabilitiesFunc;
+            readonly IAsyncClientCapabilitiesServiceV2 inner;
+            readonly GetCapabilitiesClientDecorator getCapabilitiesFunc;
+            readonly Func<IAsyncClientCapabilitiesServiceV2, Task> beforeGetCapabilities;
+            readonly Func<CapabilitiesResponseV2?, Task> afterGetCapabilities;
 
-            public FuncCapabilitiesServiceV2Decorator(IAsyncClientCapabilitiesServiceV2 inner, GetCapabilitiesClientDecorator getCapabilitiesFunc)
+            public FuncCapabilitiesServiceV2Decorator(
+                IAsyncClientCapabilitiesServiceV2 inner,
+                GetCapabilitiesClientDecorator getCapabilitiesFunc,
+                Func<IAsyncClientCapabilitiesServiceV2, Task> beforeGetCapabilities,
+                Func<CapabilitiesResponseV2?, Task> afterGetCapabilities
+            )
             {
                 this.inner = inner;
                 this.getCapabilitiesFunc = getCapabilitiesFunc;
+                this.beforeGetCapabilities = beforeGetCapabilities;
+                this.afterGetCapabilities = afterGetCapabilities;
             }
 
             public async Task<CapabilitiesResponseV2> GetCapabilitiesAsync(HalibutProxyRequestOptions options)
             {
-                return await getCapabilitiesFunc(inner, options);
+                CapabilitiesResponseV2? response = null;
+                try
+                {
+                    await beforeGetCapabilities(inner);
+                    response = await getCapabilitiesFunc(inner, options);
+                    return response;
+                }
+                finally
+                {
+                    await afterGetCapabilities(response);
+                }
```

Comment:
The various script service decorators didn't need to do this as a new builder is created for before and after actions, rather than the shared builder approach used here.